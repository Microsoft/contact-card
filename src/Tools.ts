import * as React from "react";
import { Button, BaseButton, LinkBase } from "office-ui-fabric-react";

export function openLink(url: string, e?: React.SyntheticEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement | HTMLDivElement | BaseButton | Button | LinkBase>) {
    let target: string | undefined = "_self";   // open in same window by default. This also prevents empty tabs for `mailto:` links in Chrome

    if (window.self !== window.top) {           // CSP (Content Security Policy) can prevent URL changing if we're inside iFrame. We let browser to decide where to open the URL
        target = undefined;
    }

    window.open(url, target);

    if (e) {
        e.preventDefault();
        e.stopPropagation();
    }
}
