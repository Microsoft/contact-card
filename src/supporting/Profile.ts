import { IPersonaProfile } from "../Types";


export function buildProfileResponse(id: number) {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        mail: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhones: [`555-123-456${id}`],
        imAddresses: [`im-user${id}@contoso.com`],
        companyName: "Contoso"
    };
}


export function buildProfile(id: number): IPersonaProfile {
    return {
        id: `userId${id}`,
        displayName: `Display Name ${id}`,
        jobTitle: `Job Title ${id}`,
        department: `Department ${id}`,
        email: `user${id}@contoso.com`,
        officeLocation: `1234-${id}`,
        city: "Redmond",
        businessPhone: `555-123-456${id}`,
        imAddress: `im-user${id}@contoso.com`,
        companyName: "Contoso"
    };
}
