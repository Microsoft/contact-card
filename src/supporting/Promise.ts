export async function delay(ms: number) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function waitNextTick() {
    return delay(0);
}
