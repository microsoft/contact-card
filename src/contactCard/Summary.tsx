import * as React from "react";
import { ActionButton, Icon, Link, PersonaSize, Shimmer, ShimmerElementType } from "@fluentui/react";
import { Persona } from "../Persona";
import { IPersonaProfile, PersonaShowMode } from "../Types";
import { openLink } from "../Tools";


export function renderSummary(
    profile: IPersonaProfile,
    manager: IPersonaProfile | undefined,
    isManagerLoading: boolean,
    onContactDetailsClick: () => void,
    onOrgDetailsClick: () => void,
    onPersonaClick: (profile: IPersonaProfile) => void
): React.ReactNode {
    return (
        <ul tabIndex={-1} className="summary">
            {renderContactSummary(profile, onContactDetailsClick)}
            {renderOrgSummary(manager, isManagerLoading, onOrgDetailsClick, onPersonaClick)}
        </ul>
    ); 
}

    let isExpanded = false;
    function handleContactDetailsClick(onContactDetailsClick: () => void) {
        onContactDetailsClick();
        const button = document.querySelector('[data-focus="button"]');
        if (button) {
          (button as HTMLButtonElement).focus();
        }
        isExpanded = true;
    }
  
function renderContactSummary(profile: IPersonaProfile, onContactDetailsClick: () => void): React.ReactNode {
    return (
        <li>
            <ActionButton className="section-title contact-details-button" onClick={onContactDetailsClick}>
                Contact <Icon iconName="ChevronRight" className="chevron-icon" />
            </ActionButton>

            <div className="contact-row">
                <Icon iconName="Mail" className="contact-icon" />
                <Link href={`mailto:${profile.email}`} className="contact-link email" onClick={e => openLink(`mailto:${profile.email}`, e)} aria-label={`Email ${profile.email}`}>{profile.email}</Link>
            </div>
            {
                profile.businessPhone &&
                <div className="contact-row">
                    <Icon iconName="Phone" className="contact-icon" />
                    <Link href={`tel:${profile.businessPhone}`} className="contact-link business-phone" onClick={e => openLink(`tel:${profile.businessPhone}`, e)} aria-label={`Phone ${profile.businessPhone}`}>{profile.businessPhone}</Link>
                </div>
            }
            <div className="contact-row">
                <Icon iconName="POI" className="contact-icon" />
                <span>{profile.officeLocation}</span>
                <span>&nbsp;{profile.city}</span>
            </div>
            <ActionButton className="more-details contact-details" onClick={() => handleContactDetailsClick(onContactDetailsClick)} aria-label={isExpanded ? 'show more button expanded' : 'show more button collapsed'} aria-expanded={isExpanded} data-focus="button" aria-live="polite">
            {isExpanded ? "expanded" : "collapsed"}
                Show more
            </ActionButton>
        </li>
    );
}


function renderOrgSummary(
    manager: IPersonaProfile | undefined,
    isManagerLoading: boolean,
    onOrgDetailsClick: () => void,
    onPersonaClick: (profile: IPersonaProfile) => void
): React.ReactNode {
    return (
        <li>
            {(manager || isManagerLoading) &&
                <>
                    <ActionButton className="section-title org-details-button" onClick={onOrgDetailsClick}>
                        Reports to <Icon iconName="ChevronRight" className="chevron-icon" />
                    </ActionButton>
                    {
                        manager ?
                            <ActionButton className="person manager" onClick={() => onPersonaClick(manager)}>
                                <Persona
                                    id={manager.id}
                                    displayName={manager.displayName}
                                    showMode={PersonaShowMode.NameTitle}
                                    size={PersonaSize.size40}
                                />
                            </ActionButton>
                            :
                            <div className="person">
                                <Shimmer shimmerElements={[{ type: ShimmerElementType.circle, height: 40 }, { type: ShimmerElementType.gap, width: 12 }, { type: ShimmerElementType.line }]} width={"80%"} />
                            </div>
                    }
                </>
            }
            <ActionButton className="more-details org-details" onClick={onOrgDetailsClick}>
                Show organization
            </ActionButton>
        </li>
    );
}
