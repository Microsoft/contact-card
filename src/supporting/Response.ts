import { buildProfileResponse } from "./Profile";

export function buildResponse(userId: number, status: number) {
    return new Response(JSON.stringify({
        responses: [
            {
                id: userId,
                body: buildProfileResponse(userId),
                status: status
            }
        ]
    }));
}
