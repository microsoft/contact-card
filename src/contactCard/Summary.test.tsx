import * as React from "react";
import * as Enzyme from "enzyme";
import { renderSummary } from "./Summary";
import { buildProfile } from "../supporting/Profile";

jest.mock("../Tools");
import { openLink } from "../Tools";


test("renders full Summary", () => {
    const wrap = Enzyme.shallow(<div>{renderSummary(buildProfile(1), buildProfile(2), false, jest.fn(), jest.fn(), jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("renders Summary with loading manager", () => {
    const wrap = Enzyme.shallow(<div>{renderSummary(buildProfile(1), undefined, true, jest.fn(), jest.fn(), jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("renders Summary w/o any manager", () => {
    const wrap = Enzyme.shallow(<div>{renderSummary(buildProfile(1), undefined, false, jest.fn(), jest.fn(), jest.fn())}</div>);
    expect(wrap).toMatchSnapshot();
});


test("Summary handles clicks", () => {
    const onContactDetailsClick = jest.fn();
    const onOrgDetailsClick = jest.fn();
    const onPersonaClick = jest.fn();
    const wrap = Enzyme.shallow(<div>{renderSummary(buildProfile(1), buildProfile(2), false, onContactDetailsClick, onOrgDetailsClick, onPersonaClick)}</div>);

    wrap.find(".contact-details-button").simulate("click");
    expect(onContactDetailsClick).toBeCalledTimes(1);

    onContactDetailsClick.mockReset();
    wrap.find(".more-details.contact-details").simulate("click");
    expect(onContactDetailsClick).toBeCalledTimes(1);

    wrap.find(".org-details-button").simulate("click");
    expect(onOrgDetailsClick).toBeCalledTimes(1);

    onOrgDetailsClick.mockReset();
    wrap.find(".more-details.org-details").simulate("click");
    expect(onOrgDetailsClick).toBeCalledTimes(1);

    wrap.find(".person.manager").simulate("click");
    expect(onPersonaClick).toBeCalledTimes(1);
    expect(onPersonaClick.mock.calls[0][0]).toEqual(buildProfile(2));
});


test("Summary opens links", () => {
    const wrap = Enzyme.shallow(<div>{renderSummary(buildProfile(1), buildProfile(2), false, jest.fn(), jest.fn(), jest.fn())}</div>);
    wrap.find(".contact-link.email").simulate("click");
    wrap.find(".contact-link.business-phone").simulate("click");
    expect(openLink).toBeCalledTimes(2);
});
