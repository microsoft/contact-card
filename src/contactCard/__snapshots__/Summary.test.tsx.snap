// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Summary w/o any manager 1`] = `
<div>
  <ul
    className="summary"
    tabIndex={-1}
  >
    <li>
      <CustomizedActionButton
        className="section-title contact-details-button"
        onClick={[MockFunction]}
      >
        Contact 
        <Icon
          className="chevron-icon"
          iconName="ChevronRight"
        />
      </CustomizedActionButton>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Mail"
        />
        <StyledLinkBase
          aria-label="Email user1@contoso.com"
          className="contact-link email"
          href="mailto:user1@contoso.com"
          onClick={[Function]}
        >
          user1@contoso.com
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Phone"
        />
        <StyledLinkBase
          aria-label="Phone 555-123-4561"
          className="contact-link business-phone"
          href="tel:555-123-4561"
          onClick={[Function]}
        >
          555-123-4561
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="POI"
        />
        <span>
          1234-1
        </span>
        <span>
           
          Redmond
        </span>
      </div>
      <CustomizedActionButton
        aria-expanded={true}
        aria-label="Show more"
        className="more-details contact-details"
        onClick={[MockFunction]}
        tabIndex={0}
      >
        Show more
      </CustomizedActionButton>
    </li>
    <li>
      <CustomizedActionButton
        className="more-details org-details"
        onClick={[MockFunction]}
      >
        Show organization
      </CustomizedActionButton>
    </li>
  </ul>
</div>
`;

exports[`renders Summary with loading manager 1`] = `
<div>
  <ul
    className="summary"
    tabIndex={-1}
  >
    <li>
      <CustomizedActionButton
        className="section-title contact-details-button"
        onClick={[MockFunction]}
      >
        Contact 
        <Icon
          className="chevron-icon"
          iconName="ChevronRight"
        />
      </CustomizedActionButton>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Mail"
        />
        <StyledLinkBase
          aria-label="Email user1@contoso.com"
          className="contact-link email"
          href="mailto:user1@contoso.com"
          onClick={[Function]}
        >
          user1@contoso.com
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Phone"
        />
        <StyledLinkBase
          aria-label="Phone 555-123-4561"
          className="contact-link business-phone"
          href="tel:555-123-4561"
          onClick={[Function]}
        >
          555-123-4561
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="POI"
        />
        <span>
          1234-1
        </span>
        <span>
           
          Redmond
        </span>
      </div>
      <CustomizedActionButton
        aria-expanded={true}
        aria-label="Show more"
        className="more-details contact-details"
        onClick={[MockFunction]}
        tabIndex={0}
      >
        Show more
      </CustomizedActionButton>
    </li>
    <li>
      <CustomizedActionButton
        className="section-title org-details-button"
        onClick={[MockFunction]}
      >
        Reports to 
        <Icon
          className="chevron-icon"
          iconName="ChevronRight"
        />
      </CustomizedActionButton>
      <div
        className="person"
      >
        <StyledShimmerBase
          shimmerElements={
            Array [
              Object {
                "height": 40,
                "type": 2,
              },
              Object {
                "type": 3,
                "width": 12,
              },
              Object {
                "type": 1,
              },
            ]
          }
          width="80%"
        />
      </div>
      <CustomizedActionButton
        className="more-details org-details"
        onClick={[MockFunction]}
      >
        Show organization
      </CustomizedActionButton>
    </li>
  </ul>
</div>
`;

exports[`renders full Summary 1`] = `
<div>
  <ul
    className="summary"
    tabIndex={-1}
  >
    <li>
      <CustomizedActionButton
        className="section-title contact-details-button"
        onClick={[MockFunction]}
      >
        Contact 
        <Icon
          className="chevron-icon"
          iconName="ChevronRight"
        />
      </CustomizedActionButton>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Mail"
        />
        <StyledLinkBase
          aria-label="Email user1@contoso.com"
          className="contact-link email"
          href="mailto:user1@contoso.com"
          onClick={[Function]}
        >
          user1@contoso.com
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="Phone"
        />
        <StyledLinkBase
          aria-label="Phone 555-123-4561"
          className="contact-link business-phone"
          href="tel:555-123-4561"
          onClick={[Function]}
        >
          555-123-4561
        </StyledLinkBase>
      </div>
      <div
        className="contact-row"
      >
        <Icon
          className="contact-icon"
          iconName="POI"
        />
        <span>
          1234-1
        </span>
        <span>
           
          Redmond
        </span>
      </div>
      <CustomizedActionButton
        aria-expanded={true}
        aria-label="Show more"
        className="more-details contact-details"
        onClick={[MockFunction]}
        tabIndex={0}
      >
        Show more
      </CustomizedActionButton>
    </li>
    <li>
      <CustomizedActionButton
        className="section-title org-details-button"
        onClick={[MockFunction]}
      >
        Reports to 
        <Icon
          className="chevron-icon"
          iconName="ChevronRight"
        />
      </CustomizedActionButton>
      <CustomizedActionButton
        className="person manager"
        onClick={[Function]}
      >
        <Persona
          displayName="Display Name 2"
          id="userId2"
          showMode={3}
          size={12}
        />
      </CustomizedActionButton>
      <CustomizedActionButton
        className="more-details org-details"
        onClick={[MockFunction]}
      >
        Show organization
      </CustomizedActionButton>
    </li>
  </ul>
</div>
`;
